diff --git a/Android.mk b/Android.mk
index b8eaadd..5903607 100755
--- a/Android.mk
+++ b/Android.mk
@@ -32,6 +32,13 @@ LOCAL_MULTILIB := 32
 
 LOCAL_PROGUARD_ENABLED := disabled
 
+ifeq ($(strip $(BOARD_HAVE_FMRADIO_BCM)),true)
+LOCAL_SRC_FILES += \
+        $(call all-java-files-under, fm/app/src)
+LOCAL_STATIC_JAVA_LIBRARIES += com.broadcom.fm
+LOCAL_FULL_MANIFEST_FILE := $(LOCAL_PATH)/fm/app/AndroidManifest.xml
+endif
+
 include $(BUILD_PACKAGE)
 
 include $(call all-makefiles-under,$(LOCAL_PATH))
diff --git a/jni/Android.mk b/jni/Android.mk
index 581b290..82ca9ac 100644
--- a/jni/Android.mk
+++ b/jni/Android.mk
@@ -17,6 +17,13 @@ LOCAL_SRC_FILES:= \
     com_android_bluetooth_sdp.cpp \
     com_android_bluetooth_btservice_vendor.cpp
 
+
+ifeq ($(strip $(BOARD_HAVE_FMRADIO_BCM)),true)
+LOCAL_SRC_FILES += ../fm/app/jni/com_broadcom_fm_service.cpp
+LOCAL_CFLAGS += -DBOARD_HAVE_FMRADIO_BCM
+endif
+
+
 ifneq ($(TARGET_SUPPORTS_WEARABLES),true)
 LOCAL_C_INCLUDES += \
      $(JNI_H_INCLUDE) \
diff --git a/jni/com_android_bluetooth.h b/jni/com_android_bluetooth.h
index a797dd3..3662d46 100644
--- a/jni/com_android_bluetooth.h
+++ b/jni/com_android_bluetooth.h
@@ -54,6 +54,10 @@ int register_com_android_bluetooth_gatt (JNIEnv* env);
 
 int register_com_android_bluetooth_sdp (JNIEnv* env);
 
+#ifdef BOARD_HAVE_FMRADIO_BCM
+int register_com_broadcom_fm_service(JNIEnv* env);
+#endif
+
 int register_com_android_bluetooth_btservice_vendor (JNIEnv* env);
 }
 
diff --git a/jni/com_android_bluetooth_btservice_AdapterService.cpp b/jni/com_android_bluetooth_btservice_AdapterService.cpp
index 95cc664..432d448 100644
--- a/jni/com_android_bluetooth_btservice_AdapterService.cpp
+++ b/jni/com_android_bluetooth_btservice_AdapterService.cpp
@@ -1563,10 +1563,26 @@ jint JNI_OnLoad(JavaVM *jvm, void *reserved)
         return JNI_ERR;
     }
 
+#ifdef BOARD_HAVE_FMRADIO_BCM
+    if ((status = android::register_com_broadcom_fm_service(e)) < 0) {
+        ALOGE("jni fm registration failure: %d", status);
+        return JNI_ERR;
+    }
+#endif
+
+
+
     if ((status = android::register_com_android_bluetooth_btservice_vendor(e)) < 0) {
         ALOGE("jni vendor registration failure: %d", status);
         return JNI_ERR;
     }
 
+//#ifdef BOARD_HAVE_FMRADIO_BCM
+//    if ((status = android::register_com_broadcom_fm_service(e)) < 0) {
+//        ALOGE("jni fm registration failure: %d", status);
+//        return JNI_ERR;
+//    }
+//#endif
+
     return JNI_VERSION_1_6;
 }
diff --git a/src/com/android/bluetooth/btservice/AdapterService.java b/src/com/android/bluetooth/btservice/AdapterService.java
index b6caeb2..ad5c1f4 100644
--- a/src/com/android/bluetooth/btservice/AdapterService.java
+++ b/src/com/android/bluetooth/btservice/AdapterService.java
@@ -551,13 +551,13 @@ public class AdapterService extends Service {
     }
     public boolean onUnbind(Intent intent) {
         Log.w(TAG, "onUnbind, calling cleanup");
-        cleanup();
         return super.onUnbind(intent);
     }
 
     public void onDestroy() {
         debugLog("onDestroy()");
         mProfileObserver.stop();
+        cleanup();
     }
 
     void BleOnProcessStart() {
