diff --git a/src/com/android/nfc/NfcService.java b/src/com/android/nfc/NfcService.java
old mode 100755
new mode 100644
index 2c0e7c7..8781878
--- a/src/com/android/nfc/NfcService.java
+++ b/src/com/android/nfc/NfcService.java
@@ -553,7 +553,7 @@ public class NfcService implements DeviceHostListener {
                         enableInternal();
                     } else {
                         Log.d(TAG, "NFC is off.  Checking firmware version");
-                        mDeviceHost.checkFirmware();
+                        //mDeviceHost.checkFirmware();
                     }
                     if (mPrefs.getBoolean(PREF_FIRST_BOOT, true)) {
                         Log.i(TAG, "First Boot");
